### ISSUE TITLE
bigger stdin transfer unit

### ISSUE THREAD

**Issue by wangyugui** [open-mpi/ompi#6162]
dd if=/dev/zero bs=1M count=2 2>/dev/null | mpirun --allow-run-as-root -np 1 dd bs=1M of=a.txt
0+291 records in
0+291 records out

then we can see this 1M*2 of stdin are transfered by 291 times by about 7K unit.

openmpi version:3.1.3 4.0.0

can we use bigger stdin transfer unit then better performance?

the pipeline buffer is 8K, and O_NONBLOCK used inside openmpi stdin transfer.

OR just standard input file handle inherits without stdin transfer when rank 0 and mpirun are in same node?

Best Regards.

**Comment by ggouaillardet** [comment]
@wangyugui Using the `--allow-run-as-root` option and reading from `stdin` is considered as bad practice.

You should instead run with a non root account, and update your application to read from a file (whose name can be passed on the command line or via a config file). If you implement the latter change, you will seamlessly and efficiently support direct launch (such as `srun` or `prun`).

I suggested such optimization when rank 0 and mpirun are running on the same node, but it was rejected because a) we do not want to encourage endusers to read from the command line, and b) we want to keep the code simple.

Bottom line, it is very unlikely `stdin` (and I guess writing to `stdout` or `stderr` also use small transfer units too) transfer unit will be changed for enhanced performances.

**Comment by wangyugui** [comment]
Thanks for your reply.

We read from the stdin for MPI application in this case.
```
 data_process_1_local_application (read from local system)|
 data_process_2_MPI_application |
 data_process_3_local_application (save the result to local file system )
```
In this case, we need not to save the temp result file.

I will check/test prun.

**Comment by ggouaillardet** [comment]
You can used named pipes https://linux.die.net/man/1/mkfifo but this is not something I recommend. You could instead make the pre/post processing MPI and then
```
mpirun -np 1 preprocessor : -np n compute : -np 1 opal_process
```
And then send MPI messages at full speed between your apps.

**Comment by wangyugui** [comment]
Yes. mkfifo can work out this problem. but a little complex usage.

```
mkfifo fifo_pipe

data_process_1_local_application &  (read from local file system , and write to fifo_pipe),
sleep 5 to make sure fifo is open for write firstly.
data_process_2_MPI_application  to open fifo to read.

```

**Comment by jsquyres** [comment]
Can I ask what your overall goal is here?

Reading from stdin is quite definitely not intended to be a high performance channel in Open MPI.

It is much more efficient to read from files; either directly in each MPI process, or -- as @ggouaillardet suggests -- reading them in a single MPI process and then using high-performance MPI APIs to send the data from one process to another (or multiple others).

**Comment by wangyugui** [comment]
Our goal  is to make our MPI application to work more nice/better performance for this case.

```
 data_process_1_local_application (read from local file system ) |
 data_process_2_MPI_application |
 data_process_3_local_application (save the result to local file system )

work as a pipeline. 
No needs to save the temp file.
```

For this case, we need better stdin  support.
```
dd if=/dev/zero bs=1M count=2 2>/dev/null | mpirun --allow-run-as-root -np 1 dd bs=1M of=a.txt
```
show that Reading from stdin in  openmpi now is not well designed for performance.

Can we do some performance upgrade for this case?

method 1:
now Read from stdin with O_NONBLOCK (call read() once?), and then send to other MPI process.
can we Read from stdin with O_NONBLOCK until no data ready(call reads() once or multiple times), merge it to one bigger data, and then send to other MPI process?
this will result in better performance?

method 2:
some optimization when rank 0 and mpirun are running on the same node, such as by fork, not by ssh to start MPI process?

**Comment by ggouaillardet** [comment]
@wangyugui This is a bit of a trick, but here is yet an other option.
You can start your MPI app in singleton mode (e.g. `a.out` instead of `mpirun -np <n> a.out`) and then have it `MPI_Comm_spawn()` the other tasks). If you do so, I do believe rank 0 will directly read/write from/to the shell pipe rather than `mpirun`.

**Comment by wangyugui** [comment]
In singleton mode,  we will have some feature of mpirun inside in our application. such as 
1) mpirun --host support(must)
2) mpirun socket/numa binding(option)
If so , it seems not a good idea.

**Comment by ggouaillardet** [comment]
you can pass a host/hostfile to `MPI_Comm_spawn()` via a `MPI_Info`, and specify binding options via environment variables.

**Comment by ggouaillardet** [comment]
and by the way, I do not understand why you want to open the named pipe in write mode __before__ you open it in read mode.

**Comment by wangyugui** [comment]
@ggouaillardet 

Thanks a lot for this  idea based on MPI_Comm_spawn(). 
It maybe a good choice to support current version of both OpenMPI and MPICH.

But starting process by fork rather than rsh when rank 0 and mpirun are running on the same node maybe a final solution in the further?

**Comment by wangyugui** [comment]
@ggouaillardet 

No needs of firstly opening it for writing for named pipe(mkfifo). That was my mistake.

**Comment by wangyugui** [comment]
mkfifo is a good enough way to walk around this problem now. so close this ticket.
